tactics/tactics.ml

```diff
@@ -354,7 +354,7 @@ let convert_concl ~cast ~check ty k =
         end else sigma
       in
       let (sigma, x) = Evarutil.new_evar env sigma ~principal:true ty in
-      let ans = if not cast then x else mkCast(x,k,conclty) in
+      let ans = if not cast || EConstr.eq_constr sigma ty conclty then x else mkCast(x,k,conclty) in
```

Comment:
Term equality is very broken, and to add insult to injury this is O(n). Can't we get a more reasonable API for reduction functions that would signal when nothing was changed?

Reply:
If a more reasonable API is too long-term of a project, could it make sense to test for physical equality here? I suspect (but haven't checked) that some reduction strategies would actually leave the term untouched if they perform no work.

Reply:
Most reductions touch the term by unzipping the stack then re-zipping even when nothing is done AFAIK

Reply:
Hm. So this re-zipping could be avoided if the reduction functions tracked their own progress and returned a signal as suggested above? I imagine the re-zipping itself is fast but using the original term would have benefits in every place that has a fast path for physical equality, wouldn't it?

Reply:
One has to be very careful, notably regarding evar and universe substitution. It might be the case that an evar is expanded by the reduction leading to an actually different term, but "equal" to it, although not physically. Regardless of the considered equality, I really believe that this PR is wandering on dangerous grounds and prone to introduce hard to debug issues.

Reply:
Stupid Q but is `progress` reliable? How does this compare with `try progress simpl`?

> I imagine the re-zipping itself is fast but using the original term would have benefits in every place that has a fast path for physical equality, wouldn't it?

It also limits CPU cache footprint